Fixes various issues with ONS RPC

jagerman opened the pull request:
- add a cmake option,` -DWITH_NLOHMANN_JSON_DIAGNOSTICS=ON`, to enable nlohmann::json extended diagnostics to make working with json exceptions much easier.  The option defaults to ON for non-Release builds.
- ons_names_to_owners in the wallet was calling the entirely wrong daemon endpoint
- get_mappings was returning nothing when given no types, instead of all types.
- replaced ONS_NAMES_TO_OWNERS awkward interfaces with a new, simpler ONS_INFO
- fix some bugs breaking simplewallet's ons_by_owner command

Doy-lee committed:
Fix ons_names_to_owners executing get_output_histogram

Doy-lee committed:
Fix ons_buy_mapping and ons_lookup reading wrong params

- Wallet code for generating ONS_NAMES_TO_OWNERS was creating a

  {"entries": ["name_hash": ..., "type": ...}}

  Payload which is not the correct form. This endpoint now expects

  {"name_hash": [...], "types": [...]}

  So update all the wallet code to produce the correct queries.

- Similarly the response expected an array of 'entries' which no longer
  exists. The RPC endpoint responds with something along the lines of

  {"result": {...}, "status": "OK"}

  Wallet code has been updated to read from the 'result' object.

- The old end point used to return an "entry index" which was the name
  to which a decrypted value was associated with. This way, it's possible
  to determine which names didn't exist/failed to be decrypted by matching
  the index in the response with the request. This was removed, breaking this
  feature and instead requiring a decryption of the result by iterating through
  all the names we requested.

  In terms of these APIs, it'd be better to return the nil object (zero
  initialised responses) for each name that it failed to decrypt. This
  makes the usage code trivial to handle the result for each name we
  requested, there's a 1:1 relationship between the number of input
  arguments and output objects.

  Choosing to have optional fields and omitting them as we have on the
  RPC server or having to match up keys in the request and response
  output in this case pushes the processing complexity to the caller which
  unnecessarily bleeds implementation detail into the calling code.

- Fix a crash in the core rpc server when an ONS is requested via this endpoint
  from array OOB access (where the code expects a ONS mapping to be
  returned irrespective of if it exists or not).

jagerman committed:
Fix ONS get_mapings returning nothing when types is empty

Empty is supposed to mean all types, and the later code here actually
handles that (by not constraining the query when types is empty), but
the early return here meant that with no types you got nothing.

Also clean up the placeholder string builder slightly using fmt.

jagerman committed:
Add cmake option to enable nlohmann::json diagnostics

Turning this on allows json exceptions to include the element name that
caused the exception, e.g.

    [json.exception.type_error.302] (/address/housenumber) type must be number, but is string

instead of:

    [json.exception.type_error.302] type must be number, but is string

It does, however, slightly increase the size of every json object, json
exceptions, and incurs some runtime overhead, so this is disabled by
default for Release builds.

jagerman committed:
Replace ONS_NAMES_TO_OWNERS with new ONS_INFO

ONS_NAMES_TO_OWNERS originally had a difficult interface because it was
being used to do too much (both multiple looks + multiple types within a
lookup), and the Oxen 11 revamp of it both broke the old and and left it
still difficult to use.

This drops it entirely and replaces it with a ONS_INFO that accomplishes
the same thing, but vastly simplifies it by:

- only taking one name_hash at a time
- not taking type (it just returns all found types)

This is slightly limited, but it's rare that we do multiple name lookups
and once, and in the rare case where it's needed, multiple RPC calls
isn't a big deal, and this cuts out a lot of code on both ends of the
RPC request.

It also restores the ability to query expired entries, which got lost in
the Oxen 11 RPC refactor.

Also adds documentation for the (new) method.

jagerman committed:
Fix ons-by-owner RPC issues

- wallet wasn't handling a null value for some fields, like backup_owner
- the null value should just have been omitted anyway
- fixed binary values (encrypted value, txid) not being binary when
  invoked as a bt-encoded request.
- uncovered the mystery of how assigning a vector of random structs was
  becoming json (there was a `to_json` doing the heavy lifting).  I
  replaced it, however, as the to_json was also responsible for the
  nulls and the lack of binary values.

Doy-lee commented on external/CMakeLists.txt:
```diff
@@ -145,6 +145,22 @@ if(BUILD_PYBIND)
 endif()
 
 
+if(NOT TARGET nlohmann_json::nlohmann_json)
```
For this to build for me where I want to use the vendored nlohmann::json, it wants to be put under ethyl, since ethyl defines the target for me and has the source code. But having a look at the core build, it seems like the only place we're getting nlohmann::json from is ethyl?

In terms of library design, it's a little bit odd that we export nlohmann::json publically from the ethyl target because that forces downstream projects to adopt nlohmann::json for functions in the interface that should really return you the raw JSON payload (and let projects that integrate it parse the JSON using whichever method they like).

Anyway this is all to say that we don't currently vendor nlohmann in core so this branch will execute and look for a submodule that doesn't exist (if you didn't have it installed via the system). Moving it below ethyl works but ethyl is not `JSON_DIAGNOSTICS` configurable so hence my statement about the inflexible design we have got setup right now that at the outset I'd address by fixing ethyl to not export nlohmann.

jagerman replied:
Ah, I had a `external/nlohmann-json` checkout sitting in my source dir, and I didn't realize that was some old submodule clone and not part of the current dev branch.

jagerman replied:
I do agree with the bit about ethyl and nlohmann::json, but that's a bigger job than is feasible by the next release, and so for now this has just updated to load nlohmann::json out of ethyl's submodule, but before ethyl so that we can still add diagnostics to the target cleanly.  (Ethyl will just use a predefined nlohmann_json::nlohmann_json if one exists).

Doy-lee commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -2889,6 +2869,49 @@ struct ONS_RESOLVE : PUBLIC {
+    static constexpr size_t MAX_REQUEST_ENTRIES = 256;
+
+    struct request_parameters {
+        std::string name_hash;
```
I'd prefer this to be an array, base API takes an array, helper functions abstract ontop of bulk API to do N=1. Apps can do bulk operations with minimal overhead.

If so, then `MAX_REQUEST_ENTRIES` will need to be used again in the RPC handler, or otherwise deleted from the interface.

jagerman replied:
> I'd prefer this to be an array, base API takes an array, helper functions abstract ontop of bulk API to do N=1. Apps can do bulk operations with minimal overhead.

I don't think I follow this.  If by base API you mean the internal API, it actually doesn't take an array; the old version of core_rpc_server.cpp simply looped through all the inputs and made separate internal API calls to obtain the names for each input.

At the same time, an array input here *greatly* increased the complexity of returning and dealing with the returned result in RPC-API-consuming code, for a use case that is used basically nowhere.  So it increases RPC API complexity, increases RPC implementation complexity, all for a use case that no one uses, where it isn't a big deal to simply make multiple requests.

Should that need arise, however, a generic "batch" rpc endpoint (as storage server provides, for example) where one request can contain multiple subrequests would be a much cleaner way to go about this, both for this and other endpoints.

> or otherwise deleted from the interface.

My bad, that was meant to be deleted.

jagerman replied:
I've amended this some more, to bring back the ability to request multiple at once, but keeping the interface a little less awkward to use, as the RPC wallet has a call (actively used in the GUI wallet) that required either this, or adding a generic batch method.  You now can pass an array of name hashes and get:

```json
    "result": {
        "abc": [{"type": 0, ...}, {"type": 1, ...}],
        "xyz": []
    }
```
where each key corresponds to a provided input `"name_hash"` value (and the actual input, *not*, as was previously the case, the hash reconverted to base64).

Doy-lee commented on src/wallet/wallet2.cpp:
```diff
@@ -9822,81 +9822,75 @@ static ons_prepared_args prepare_tx_extra_oxen_name_system_values(
-        if (!success) {
+    nlohmann::json req_params{
+            {"name_hash", std::array{oxenc::to_base64(tools::view_guts(result.name_hash))}},
+            {"types", std::vector<uint16_t>{ons::db_mapping_type(type)}}};
```
As it stands, `ons_info` takes 1 item, so this needs to be changed. If `ons_info` turns into a bulk API then this can remain unchanged.

Types is no longer accepted, this is going to return all the mappings for the hash so the body of this logic will have to change to lookup the type that's returned for the particular operation we're trying to perform on the ONS record.

jagerman replied:
Yeah, `type` is indeed needed, and I've re-added it and updated this code to de-array the arguments.

jagerman committed:
Fix path to nlohmann-json submodule

jagerman committed:
drop obsolete variable

jagerman committed:
Add `type` filter to ons_info

Some endpoints (such as the one fixed here) only care about a single
type and we can/should easily filter for a single type.

jagerman committed:
Improve ONS name check error messages

The error messages (which get displayed to users in wallets) were quite
raw; this makes them more human-readable.

It also fixes a bug where a check for an invalid session name ending
with an invalid name (such as "foo[") printed an error that names cannot
end with hyphens.  This was happening because the last-char check being
done *before* the overall validity check, and so we hadn't already
checked that the name only has a-z0-9_- at that point.  Rearranging the
order of the checks fixes it.

jagerman committed:
wallet2: fix ons updating

This code was broken.

jagerman committed:
Mempool acceptance: only one ONS tx per name per block

We have an earlier fix in place that prevents ONS buys for the same name
in the mempool causing invalid blocks, but ONS updates and renewals can
also cause the same issue.  This removes the "is_buying()" check so that
we extend the existing filter to all ONS tx types.

jagerman committed:
ONS_INFO: support multiple lookups

It turns out that we *do* have a notable consumer of multiple lookups by
name, in the RPC wallet (with endpoint used by the GUI wallet), to do a
simultaneous lookup of all know names.  Doing this via sequential
requests would be painful for a wallet with lots of know names, so we
need a bulk version.

This updates ons_info so that the `name_hash` can be passed as either a
single value or a list, and changes the response so that instead of an
array of results, we now get a dict of results that looks like:

    "result": {
        "abc": [{"type": 0, ...}, {"type": 1, ...}],
        "xyz": []
    }

where `"abc"` and `"xyz"` are the requested name hashes (and, unlike
before, they are now exactly as input, rather than being
base64-reencoded).  This also drops the "name_hash" itself out of the
individual records (as it would be duplicated if included).

This updates various internal wallet code to return the inner record
(the `{"type":...}` above) rather than the array containing it.

Finally this updates the rpc wallet code to use this request method.

jagerman committed:
Fix error reporting of invalid txes

The `"tvc"` field is no longer included, but `"reason"` is always
included (regardless of blink status).  This fixes wallet2 to always use
the provided reason string and deletes a bunch of obsolete tvc internal
field name parsing.

jagerman committed:
Miscellaneous wallet json interaction cleanups

Fixes various suboptimal (or worse) wallet json handling.
